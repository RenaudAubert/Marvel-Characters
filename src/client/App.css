:root {
  --jumbotron-padding-y: 3rem;
}

.jumbotron {
  padding-top: var(--jumbotron-padding-y);
  padding-bottom: var(--jumbotron-padding-y);
  margin-bottom: 0;
  background-color: #fff;
}
@media (min-width: 768px) {
  .jumbotron {
    padding-top: calc(var(--jumbotron-padding-y) * 2);
    padding-bottom: calc(var(--jumbotron-padding-y) * 2);
  }
}

.jumbotron p:last-child {
  margin-bottom: 0;
}

.jumbotron-heading {
  font-weight: 300;
}

.jumbotron .container {
  max-width: 40rem;
}

footer {
  padding-top: 3rem;
  padding-bottom: 3rem;
}

footer p {
  margin-bottom: .25rem;
}

.box-shadow { box-shadow: 0 .25rem .75rem rgba(0, 0, 0, .05); }

a.custom-card,
a.custom-card:hover {
  color: inherit;
}

.card-img-left {
  height: 100%;
  border-radius: 3px 0 0 3px;
}

.flex-auto {
  -ms-flex: 0 0 auto;
  -webkit-box-flex: 0;
  flex: 0 0 auto;
}

.h-250 { height: 250px; }
@media (min-width: 768px) {
  .h-md-250 { height: 250px; }
}

.border-top { border-top: 1px solid #e5e5e5; }
.border-bottom { border-bottom: 1px solid #e5e5e5; }

#favButton, #returnButton {
padding: 0;
border: none;
background: none;
}
