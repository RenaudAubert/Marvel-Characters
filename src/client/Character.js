import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

export default class Character extends Component {
  constructor(props) {
    super(props);

    this.favClick = this.favClick.bind(this);
  }

  favClick() {
    const { character, onFavClicked } = this.props;
    onFavClicked(character.id);
  }

  // Render each character card with image, name and description
  render() {
    const { character, isFav } = this.props;
    const name = character.name || '';
    const description = character.description || 'No description available for this character';
    const imgURL = character.thumbnail && `${character.thumbnail.path}/landscape_large.${character.thumbnail.extension}`;

    return (
      <div className="h-100 card mb-4 box-shadow">
        <Link to={`/characters/${character.id}`}>
          <img className="card-img-top" src={imgURL} alt="" />
        </Link>
        <div className="card-body">
          <h5 className="card-title">{name}</h5>
          <p className={`card-text ${character.description}` ? '' : ' text-danger'}>{description}</p>
        </div>
        <div className="card-footer text-muted">
          <div className="d-flex justify-content-between align-items-center">
            <div className="btn-group">
              <Link to={`/characters/${character.id}`}>
                <button type="button" className="btn btn-sm btn-outline-secondary">
                  <i className="fas fa-search" />
                </button>
              </Link>
              <button type="button" className="btn btn-sm btn-outline-secondary" onClick={this.favClick}>
                <i className={isFav ? 'fas fa-star' : 'far fa-star'} />
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Character.propTypes = {
  character: PropTypes.shape({
    id: PropTypes.number,
    description: PropTypes.string,
    name: PropTypes.string,
    thumbnail: PropTypes.object,
  }).isRequired,
  isFav: PropTypes.bool,
  onFavClicked: PropTypes.func.isRequired
};

Character.defaultProps = {
  isFav: false
};
