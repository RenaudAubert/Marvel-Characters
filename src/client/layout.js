import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

export const FavElement = (props) => {
  const { id, name, onDeleteFav } = props;

  const deleteFav = () => {
    onDeleteFav(id);
  };

  return (
    <div className="btn-group" role="group" aria-label="Favorite List">
      <Link to={`/${id}`} className="dropdown-item btn btn-secondary text-center">
        {name}
      </Link>
      <button type="button" className="close btn btn-secondary" onClick={deleteFav}>
        <i className="fa fa-trash" aria-hidden="true" />
      </button>
    </div>
  );
};

export const Header = (props) => {
  const { favCharacters, onDeleteFav } = props;

  // for each favorite character create a favElement component
  const favList = favCharacters.map(character => (
    <FavElement
      id={character.id}
      name={character.name}
      onDeleteFav={onDeleteFav}
      key={character.id}
    />
  ));

  return (
    <header>
      <div className="navbar navbar-dark bg-dark box-shadow">
        <div className="container d-flex justify-content-between">
          <Link to="/" className="navbar-brand d-flex align-items-center">
            Marvel Heroes
          </Link>
          <div className="dropdown">
            <button className="btn btn-secondary" type="button" id="dropdownFavorite" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
              My Favorites
            </button>
            <div className="dropdown-menu" aria-labelledby="dropdownMenuButton">
              {favList.length > 0 ? favList : <p className="text-center">No favorite characters</p>}
            </div>
          </div>
        </div>
      </div>
    </header>
  );
};

export const Footer = (props) => {
  const { attributionText } = props;
  return (
    <footer className="footer text-muted">
      <div className="container">
        {attributionText}
      </div>
    </footer>
  );
};

export const FavButton = (props) => {
  const { characterId, isFav, onFavClicked } = props;

  const favClick = () => {
    onFavClicked(characterId);
  };

  return (
    <button type="button" className="btn btn-sm fa-button" onClick={favClick}>
      <i className={`${isFav ? 'fas fa-star' : 'far fa-star'} text-warning fa-lg`} />
    </button>
  );
};

// Return link if type (detail, wiki, comiclink) is found
export const LinkType = (props) => {
  const { character, type, text } = props;

  let link = text;
  const urlType = character.urls.find(url => url.type === type);
  if (urlType !== undefined) {
    link = <a href={urlType.url} target="_blank" rel="noopener noreferrer">{text}</a>;
  }
  return link;
};

FavElement.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  onDeleteFav: PropTypes.func.isRequired
};

Header.propTypes = {
  favCharacters: PropTypes.arrayOf(PropTypes.object),
  onDeleteFav: PropTypes.func.isRequired
};

Header.defaultProps = {
  favCharacters: []
};

Footer.propTypes = {
  attributionText: PropTypes.string
};

Footer.defaultProps = {
  attributionText: ''
};

FavButton.propTypes = {
  characterId: PropTypes.number.isRequired,
  isFav: PropTypes.bool,
  onFavClicked: PropTypes.func.isRequired
};

FavButton.defaultProps = {
  isFav: false
};

LinkType.propTypes = {
  character: PropTypes.shape({
    id: PropTypes.number,
    description: PropTypes.string,
    name: PropTypes.string,
    thumbnail: PropTypes.object,
  }).isRequired,
  type: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired
};

LinkType.defaultProps = {
  type: 'detail',
  text: 'detail'
};
