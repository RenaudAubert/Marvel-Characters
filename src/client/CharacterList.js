import React from 'react';
import PropTypes from 'prop-types';
import Character from './Character';

const CharacterList = (props) => {
  const { characters, onFavClicked } = props;
  let characterList = <p>No characters found</p>;

  characterList = characters.map((character) => {
    // For each character create a Character component (card in the view)
    return (
      <div className="col-md-4 pt-2" key={character.id}>
        <Character
          character={character}
          onFavClicked={onFavClicked}
        />
      </div>
    );
  });

  return (
    <main role="main">
      <div className="album py-5 bg-light">
        <div className="container">
          <div className="row">
            {characterList}
          </div>
        </div>
      </div>
    </main>
  );
};

export { CharacterList as default };

CharacterList.propTypes = {
  characters: PropTypes.arrayOf(PropTypes.object).isRequired,
  onFavClicked: PropTypes.func.isRequired,
};
