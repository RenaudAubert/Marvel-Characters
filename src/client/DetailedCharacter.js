import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { FavButton, LinkType } from './layout';

export default class DetailedCharacter extends Component {
  constructor(props) {
    super(props);

    this.state = {character: {}};
  }

  // Get data from Marvel API if component is mounted
  componentDidMount() {
    const { match, shouldFetch } = this.props;

    // Prevents fetching if character was found in characters array in App
    if (shouldFetch) {
      // Fetch a single character using id
      fetch(`/api/characters/${match.params.id}`)
      .then(res => res.json())
      .then(apiResult => {
        if (apiResult.data && apiResult.data.results.length) {
          apiResult.data.results[0].favorite = false;
          this.setState({ character: apiResult.data.results[0] });
        }
      });
    }
  }

  // Render each character card with image, name and description
  render() {
    let { character, shouldFetch, onFavClicked } = this.props;
    // Assign value returned by Marvel API if character was not found in App
    if (shouldFetch) {
      character = this.state.character;
    }
    let returnedDOM = (
      <p className="text-center text-danger">
        <strong>No character found</strong>
      </p>
    );

    // If object is empty
    if (Object.keys(character).length !== 0) {
      const imgURL = character.thumbnail ? `${character.thumbnail.path}/landscape_large.${character.thumbnail.extension}` : '';
      // Only list the 4 first comics returned
      const comicsList = (type, length) => {
        const list = character[type].items.slice(0, length).map(comic => (
          <li key={comic.name}>{comic.name}</li>
        ));

        return <ul>{list}</ul>;
      };

      returnedDOM = (
        <div>
          <div className="row">
            <div className="col-lg-5">
              <img className="card-img-left flex-auto d-none d-md-block" src={imgURL} alt="" />
            </div>
            <div className="col-lg-7">
              <h5>
                <LinkType character={character} type="detail" text={character.name} />
                <FavButton characterId={character.id} isFav={character.favorite} onFavClicked={onFavClicked} />
              </h5>
              <h6>Description</h6>
              {character.description ?
                <p className="card-text">{character.description}</p> :
                <p className="card-text text-danger">{character.description}</p>
              }
              <div className="row">
                <div className="col-sm-6">
                  <h6>
                    Comics
                    <LinkType character={character} type="comiclink" text={` (${character.comics.available})`} />
                  </h6>
                  {
                    character.comics.items.slice(0, 4).map(comic => (
                      <li key={comic.name}>{comic.name}</li>
                    ))
                  }
                  <LinkType character={character} type="comiclink" text="Details" />
                </div>
                <div className="col-sm-6">
                  <h6>Series</h6>
                  {
                    character.series.items.slice(0, 4).map(serie => (
                      <li key={serie.name}>{serie.name}</li>
                    ))
                  }
                </div>
              </div>
            </div>
          </div>
        </div>
      );
    }

    return (
      <div>
        <Link to="/" className="pl-2">
          <i className="far fa-arrow-alt-circle-left fa-3x text-secondary" />
        </Link>
        <div className="container pt-2">
          {returnedDOM}
        </div>
      </div>
  );
}
}

DetailedCharacter.propTypes = {
  character: PropTypes.shape({
    id: PropTypes.number,
    description: PropTypes.string,
    name: PropTypes.string,
    thumbnail: PropTypes.object,
    favorite: PropTypes.bool
  }),
  onFavClicked: PropTypes.func.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.node,
    }).isRequired,
  }).isRequired,
  shouldFetch: PropTypes.bool
};

DetailedCharacter.defaultProps = {
  character: {},
  shouldFetch: false
};
