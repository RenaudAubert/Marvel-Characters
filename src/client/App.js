import React, { Component } from 'react';
import { Switch, Route } from 'react-router-dom';
import DetailedCharacter from './DetailedCharacter';
import CharacterList from './CharacterList';
import Character from './Character';
import { Header, Footer } from './layout';
import './App.css';

export default class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      characters: {},
      favCharacters: []
    };

    this.handleFavClick = this.handleFavClick.bind(this);
    this.handleDeleteFav = this.handleDeleteFav.bind(this);
  }

  // Get data from Marvel API if component is mounted
  componentDidMount() {
    fetch('/api/characters')
      .then(res => res.json())
      .then(characters => this.setState({ characters }));
  }

  // Add new id into favorite array
  handleFavClick(id) {
    const { characters, favCharacters } = this.state;

    // If id already in favCharacters delete character from favorite
    if (favCharacters.findIndex(character => character.id === id) !== -1) {
      this.handleDeleteFav(id);
    } else {
      // Checks whether given id is found in characters array
      const foundCharacter = characters.data.results.find(character => id === character.id);
      if (foundCharacter) {
        this.setState(prevState => ({
          favCharacters: [...prevState.favCharacters, { id, name: foundCharacter.name }]
        }));
      }
    }
  }

  // Delete character from favorite
  handleDeleteFav(id) {
    this.setState(prevState => ({
      // Keep all elements except those matching id
      favCharacters: prevState.favCharacters.filter(character => character.id !== id)
    }));
  }

  render() {
    const { characters, favCharacters } = this.state;
    const { match } = this.props;

    // View composed of Header, Main et and Footer component
    return (
      <div>
        <Header favCharacters={favCharacters} onDeleteFav={this.handleDeleteFav} />
        <Switch>
          <Route
            exact
            path="/"
            render={props => (
              <CharacterList
                {...props}
                characters={characters}
                favCharacters={favCharacters}
                onFavClicked={this.handleFavClick}
              />
            )}
          />
          <Route
            path='/characters/:id'
            render={props => {
              return (
                <DetailedCharacter
                  {...props}
                  favCharacters={favCharacters}
                  onFavClicked={this.handleFavClick}
                />
              );
            }}
          />
        </Switch>
        <Footer attributionText={characters.attributionText} />
      </div>
    );
  }
}
