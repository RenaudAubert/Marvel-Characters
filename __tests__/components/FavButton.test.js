import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

// Component to be tested
import { FavButton } from '../../src/client/layout';

describe('<FavButton />', () => {
  const mockOnClick = jest.fn();
  const wrapper = shallow(<FavButton characterId={1} onFavClicked={mockOnClick} />);

  describe('return()', () => {
    test('renders the FavButton component', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('onClick()', () => {
    test('successfully calls the onClick handler', () => {
      wrapper.find('button').simulate('click');
      expect(mockOnClick.mock.calls.length).toEqual(1);
    });
  });
});
