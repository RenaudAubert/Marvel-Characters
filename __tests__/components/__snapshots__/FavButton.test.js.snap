// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FavButton /> return() renders the FavButton component 1`] = `
<button
  className="btn btn-sm fa-button"
  onClick={[Function]}
  type="button"
>
  <i
    className="far fa-star text-warning fa-lg"
  />
</button>
`;
